### ISSUE TITLE
Downloading all the data using Buttons extension

### ISSUE THREAD

**Issue by ldecicco-USGS** [rstudio/DT#267]
Is there a way to download all the table data using the buttons without having to display all the data?

For example, if my `rawData` table has 10,000 rows, only the default 10 are downloaded when I use the buttons.

```

    DT::datatable(rawData, extensions = 'Buttons',
                  rownames = FALSE,
                  options = list(dom = 'Bfrtip',
                                 buttons = 
                                   list('colvis', list(
                                     extend = 'collection',
                                     buttons = list(list(extend='csv',
                                                         filename = tableType),
                                                    list(extend='excel',
                                                         filename = tableType),
                                                    list(extend='pdf',
                                                         filename= tableType)),
                                     text = 'Download'
                                   )),
                                 scrollX = TRUE
                                )
    )

```

I'm aware I could create a downloadHandler, and save the data each time, but I find these buttons a really attractive option.

**Comment by joepalardy** [comment]
I was running into this issue earlier today. Are you using server-side processing?

You may want to try `DT::datatable(rawData, server = FALSE,...`

However, I suspect that with 10,000 rows you will have a performance hit, and you may want to go the downloadHandler route. The issue with DataTable buttons and server-side processing can be found on the [DataTable FAQs](https://datatables.net/faqs/#buttons).

**Comment by yihui** [comment]
Right. You have to use either `server = FALSE` or your own download button, since the upstream DataTables library does not support downloading full data in the server-side processing mode. You may file a feature request there: https://datatables.net/forums/

**Comment by cenuno** [comment]
Hi @ldecicco-USGS,

Using a combination of the `DT` and `shiny` package, I created an example showing how to download all data from the use of a datatable button. The code is down below, along with a brief explanation here, [Customizing DT Download Button](https://github.com/cenuno/shiny/tree/master/DT-Download-All-Rows-Button#customizing-dt-download-button).

I hope this helps!

Kindly,

Cristian

```R
#
# This is a Shiny web application.
#
# Find out more about building applications with Shiny here:
#
#    http://shiny.rstudio.com/
#

library(shiny)
library( DT )

# Define UI for application that creates a datatables
ui <- fluidPage(
   
   # Application title
   titlePanel("Download Datatable")
   
   # Show a plot of the generated distribution
   , mainPanel(
         DT::dataTableOutput("fancyTable")
      ) # end of main panel
   
) # end of fluid page

# Define server logic required to create datatable
server <- function(input, output) {
   
   output$fancyTable <- DT::renderDataTable(
     datatable( data = mtcars
                , extensions = 'Buttons'
                , options = list( 
                  dom = "Blfrtip"
                  , buttons = 
                    list("copy", list(
                      extend = "collection"
                      , buttons = c("csv", "excel", "pdf")
                      , text = "Download"
                    ) ) # end of buttons customization
                   
                   # customize the length menu
                  , lengthMenu = list( c(10, 20, -1) # declare values
                                       , c(10, 20, "All") # declare titles
                  ) # end of lengthMenu customization
                  , pageLength = 10
                   
                   
                ) # end of options
               
     ) # end of datatables
   )
} # end of server

# Run the application 
shinyApp(ui = ui, server = server)
```

**Comment by yihui** [comment]
@cenuno Thanks for the example! That certainly works when the data is not too big.

I thought about this issue for a couple of hours, and I think there could be a solution, but it is not trivial to implement, and requires substantial knowledge about **shiny**, **DT**, and JavaScript. This is how I would implement it:

1. Create a custom button with an `action` function: https://datatables.net/extensions/buttons/custom
1. In the `action` function, send a POST request (`$.ajax({url: URL, data: PARAMS})`), where `URL` is the Ajax URL of the table (https://datatables.net/reference/api/ajax.url()), and `PARAMS` is the parameters from the last time: https://datatables.net/reference/api/ajax.params()
1. Figure out a way to create a download from the response of the POST request using JavaScript: https://stackoverflow.com/q/16086162/559676

That is only a sketch. The real tricky part is here:
https://github.com/rstudio/DT/blob/3d8fa2479219d981c5a06f4dd31be358cc21a9aa/R/shiny.R#L370-L373

By default, the Ajax request will return the JSON data (of the current page of the table), which is not what you want. Instead, you want (1) the CSV data (or whatever format); (2) the full data. So the HTTP response from the R side needs to be changed. One way to change it is to add a special parameter in  the `PARAMS` object above (Step 2), e.g. `DT_download_data: true`. When you see this parameter on the R side (`isTRUE(params$DT_download_data)`), you prepare a special HTTP response.

Alas, this is complicated...

**Comment by cenuno** [comment]
@yihui Thank you for thinking this through! 

I'll design future Shiny apps with this in mind, where I'll label the  `shiny::downloadButton()` as "Download All Data" and I'll label the `DT` button as 'Download Some Data".

**Comment by sorhawell** [comment]
@yihui  I think I made a fairly modest workaround combining 2 stack-exchange answers. A DT button activates a modal-message wherein vanilla shiny download buttons are placed. Activating the download process imediately by pressing the DT button would be be more elegant, but this works fine.

[how-to-add-custom-button-in-r-shiny-datatable](https://stackoverflow.com/questions/53600956/how-to-add-custom-button-in-r-shiny-datatable/53608155#53608155
)
[shiny-combining-download-buttons-to-single-with-options](https://stackoverflow.com/questions/50724182/shiny-combining-download-buttons-to-single-with-options)

![image](https://user-images.githubusercontent.com/8015271/51830736-d7d3fb00-22f0-11e9-9efb-cade72c48e16.png)

```

library(shiny)
library(DT)

myModal <- function() {
  div(id = "test",
      modalDialog(downloadButton("download1","Download iris as csv"),
                  br(),
                  br(),
                  downloadButton("download2","Download iris as csv2"),
                  easyClose = TRUE, title = "Download Table")
  )
}

ui <- basicPage(
  DTOutput("dtable")
)

server <- function(input, output, session){
  output$dtable <- renderDT(
    datatable(iris,
              extensions = 'Buttons',
              options = list(
                dom = 'Bfrtip',
                buttons = list(
                  "copy",
                  list(
                    extend = "collection",
                    text = 'download entire dataset',
                    action = DT::JS("function ( e, dt, node, config ) {
                                    Shiny.setInputValue('test', true, {priority: 'event'});
}")
                  )
                    )
                  )
                )
              )
  
  observeEvent(input$test, {
    print("hello")
    showModal(myModal())
  })
  
  
  output$download1 <- downloadHandler(
    filename = function() {
      paste("data-", Sys.Date(), ".csv", sep="")
    },
    content = function(file) {
      write.csv(iris, file)
    }
  )
  
  output$download2 <- downloadHandler(
    filename = function() {
      paste("data-", Sys.Date(), ".csv", sep="")
    },
    content = function(file) {
      write.csv2(iris, file)
    }
  )
}

shinyApp(ui, server)
```

**Comment by yihui** [comment]
@sorhawell Amazing! Thanks for sharing!

**Comment by Gaurav9112** [comment]
>                   br(),
>                   br(),
>                   downloadButton("download2","Download iris as csv2"),
>                   easyClose = TRUE, title = "Download Table")

Thanks ! it worked and its much faster as compared to keeping **server = False**

**Comment by stla** [comment]
You can move a Download button inside the datatable as follows:

```
ui <- basicPage(
  downloadButton("download1","Download iris as csv"),
  DTOutput("dtable")
)

server <- function(input, output, session){
  output$dtable <- renderDT(
    datatable(iris,
              callback = JS("$('div.dwnld').append($('#download1'));"),
              extensions = 'Buttons',
              options = list(
                dom = 'B<"dwnld">frtip',
                buttons = list(
                  "copy"
                )
              )
    )
  )
  
  output$download1 <- downloadHandler(
    filename = function() {
      paste("data-", Sys.Date(), ".csv", sep="")
    },
    content = function(file) {
      write.csv(iris, file)
    }
  )
  
}
```
![Capture](https://user-images.githubusercontent.com/4466543/59583095-8c24c100-90da-11e9-9232-59c270b60271.PNG)

**Comment by stla** [comment]
This won't work if the datatable is reactive. In this case one can proceed like this:

```
library(shiny)
library(DT)

callback <- JS(
  "var a = document.createElement('a');",
  "$(a).addClass('dt-button');",
  "a.href = document.getElementById('download1').href;",
  "a.download = '';",
  "$(a).attr('target', '_blank');",
  "$(a).text('Download');",
  "$('div.dwnld').append(a);",
  "$('#download1').hide();"
)

ui <- basicPage(
  downloadButton("download1", ""), # no label: this button will be hidden
  numericInput("nrows", "Number of rows", 10),
  DTOutput("dtable")
)

server <- function(input, output, session){
  output$dtable <- renderDT(
    datatable(iris[1:input$nrows,],
              callback = callback,
              extensions = 'Buttons',
              options = list(
                dom = 'B<"dwnld">frtip',
                buttons = list(
                  "copy"
                )
              )
    )
  )
  
  output$download1 <- downloadHandler(
    filename = function() {
      paste("data-", Sys.Date(), ".csv", sep="")
    },
    content = function(file) {
      write.csv(iris, file)
    }
  )
  
}

shinyApp(ui, server)
```

**Comment by Evanleen** [comment]
Hi @stla your solution works when the name of the data table is dtable. But when the name is  changed to datatable for example (DTOutput("datatable") and output$datatable) it's no longer possible to download the file. Can you please explain how to change the name? Thank you in advance
